### PR TITLE
[dynamo] handle setting .data on a tensor

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113070
* __->__ #113069

We don't want setattr in the graph. Setting data has interesting implications on both aliasing and on the autograd engine.

The safe recipe is:

1) Disable grad
2) Call set_()
3) Manually lower the version counter on the object to hide it from the autograd engine

This is effectively the same exact thing as setting .data, and it composes properly with aot_autograd and inductor.

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/builtin.py [hunk 1]
```diff
@@ -1173,6 +1174,45 @@ def call_setattr(
...
+                    # installed in tracked fakes.
+                    to_remove = []
+                    for tf in tx.output.tracked_fakes:
+                        if tf.source == obj.source:
+                            to_remove.append(tf)
+                    for tf in to_remove:
+                        tx.output.tracked_fakes.remove(tf)
+
+                    # Step 1 - disable grad
+                    version = obj.as_proxy().node.meta["example_value"]._version
```
Comment (jon-chuang) [inline torch/_dynamo/variables/builtin.py:1194]: unused?

File: torch/_dynamo/variables/builtin.py [hunk 2]
```diff
@@ -1173,6 +1174,45 @@ def call_setattr(
...
+                                torch.Tensor.set_,
+                                *proxy_args_kwargs([obj, val], {}),
+                            ),
+                        )
+                    # Step 3 - drop the version counter - this is a hack required to get
+                    # .data setting to play correctly with the autograd engine.
+                    tx.output.create_proxy(
+                        "call_function",
+                        torch._C._autograd._unsafe_set_version_counter,
+                        (out.as_proxy(), 0),
```
Comment (jon-chuang) [inline torch/_dynamo/variables/builtin.py:1210]: Is this supposed to be 0 or version - 1?

File: torch/_dynamo/variables/builtin.py [hunk 3]
```diff
@@ -1173,6 +1174,45 @@ def call_setattr(
...
+                        {},
+                    )
+                    # Step 4 - replace all reference to the current object with the new one
+                    tx.replace_all(obj, out)
```
Comment (jon-chuang) [inline torch/_dynamo/variables/builtin.py:1214]: Although `set_` is supposed to be in-place, could we just add a test that the tensor object aliasing is preserved?

`python
z = x
x.data = y
x.add_(1)
return x is z
```

Comment (voznesenskym) [reply torch/_dynamo/variables/builtin.py:1214]: That doesn't work with aot_autograd, but it does work with eager. Talking to @bdhirsh now. See the stacked PR.

I can change this to be its own test for eager only, and that should work.

Comment (jon-chuang) [reply torch/_dynamo/variables/builtin.py:1214]: Right, I updated to just do `y.data = torch.zeros([0])`

Comment (voznesenskym) [reply torch/_dynamo/variables/builtin.py:1214]: Can add :)

File: torch/_dynamo/variables/builtin.py [hunk 4]
```diff
@@ -1173,6 +1174,45 @@ def call_setattr(
...
+                    return val.add_options(self, out, name_var)
```
Comment (jon-chuang) [inline torch/_dynamo/variables/builtin.py:1215]: I think don't return here. Setattr is supposed to return None, and we also need to track the setattr as a side effect.

Comment (jon-chuang) [reply torch/_dynamo/variables/builtin.py:1215]: So we need the code outside of this block to run

Comment (jon-chuang) [reply torch/_dynamo/variables/builtin.py:1215]: Actually, do we? Hmm. I've got a feeling it's the responsibility of AOT autograd to persist the `set_` mutation

Comment (voznesenskym) [reply torch/_dynamo/variables/builtin.py:1215]: Returning None doesn't matter here, it gets handled correctly later iirc, but we can make that change.

Object tracking can be updated a little, sure, not sure it matters tho.

Comment (jon-chuang) [reply torch/_dynamo/variables/builtin.py:1215]: Right, I think this part is fine actually. The problem is with AOT Autograd. I think tests are failing due to this: https://github.com/pytorch/pytorch/issues/110842

Comment (jon-chuang) [reply torch/_dynamo/variables/builtin.py:1215]: Let me see if I can try to handle it.

**Comment by voznesenskym** [comment]
https://github.com/pytorch/pytorch/pull/113080

**Pull request closed without merging**